### DISCUSSION TITLE
Reading camera RGB and depth buffers

### DISCUSSION THREAD

**Discussion by baryn** [vsg-dev/VulkanSceneGraph#756] [Q&A] [answered]
Hi everyone,

I have been using OSG for several years now, especially for emulating sensors like camera, depth cameras and so on. To do so, I used the RGB and depth buffers generated at each frame. I would like to move on VSG but I am wondering if it is possible to do the same. In OSG, I did something like this:

Set up camera and image
```
// create the osg::Image and allocate
auto rgb_image = new osg::Image();
rgb_image->allocateImage(cam_width, cam_height, 1, GL_RGB, GL_UNSIGNED_BYTE); // GL_DEPTH_COMPONENT for the depth buffer
// attach to the camera
 osg_camera->attach(osg::Camera::COLOR_BUFFER, rgb_image.get(),4,4);
```

Get the data
```
// in a osg::Camera::DrawCallback callback 
osg::ref_ptr<osg::Image> rgb_image;
// get the buffer size (eventually)
auto buff_size = osg_camera->rgb_img->getTotalSizeInBytes();
// get the image buffer from the camera in the scene and use it
auto rgb_buffer = rgb_image->data();
```

I had a look on the sources but could not find any clue. Is it feasible in VSG or part of further developments ?

Thank you

**Comment by timoore** [comment]
Take a look at the [vsgscreenshot example](https://github.com/vsg-dev/vsgExamples/tree/master/examples/app/vsgscreenshot) in vsgExamples; it's probably closest to what you want to do.

**Reply by baryn** [reply]
Thanks for the quick reply. I'll help !

**Comment by robertosfield** [comment, marked as answer]
Hi Bryan,

It's currently more complicated than with OpenGL/OSG, but still perfectly
possible.  Vulkan is a lower level than OpenGL so we can to explicitly
allocate staging buffer memory to copy to and then transfer the data, as
Vulkan is multi-threaded one also need to use barrieres to avoid issues.

As well as vsgscreenshot, that Tim mentioned, which does copying of the
window's depth and colour buffer, there is the vsgheadless which does the
same thing for an offscreen framebuffer.

https://github.com/vsg-dev/vsgExamples/blob/master/examples/app/vsgheadless/vsgheadless.cpp

There is common functionality between vsgscreenshot and vsgheadless which
may make sense to come up with some convenience functions for creating the
required commands.

Cheers,
Robert.

**Reply by baryn** [reply]
Hi Robert,

I understand it is indeed more complicated than with OpenGL. Thanks for pointing this example, I think I can manage from that.

Thanks for your job.
Cheers
